lib/githubbot.ts

```diff
@@ -0,0 +1,403 @@
+
+            // Request new token.
+            const tokenOpts: any = {
+                url: tokenUrl,
+                method: 'POST',
+                headers: {
+                    'Authorization': `Bearer ${jwToken}`,
+                    'Accept': 'application/vnd.github.machine-man-preview+json',
+                    'User-Agent': 'request'
+                },
+                json: true
+            };
+
+            return request.postAsync(tokenOpts);
+        }).then((res: any) => {
+            // We also need to take into account the expiry date, which will require a new kickoff.
+            const tokenDetails = res.body;
+
+            this.githubApi.authenticate({
+                type: 'token',
+                token: tokenDetails.token
+            });
+
+            // For debug.
+            this.log(ProcBot.LogLevel.DEBUG, `token for manual fiddling is: ${tokenDetails.token}`);
+            this.log(ProcBot.LogLevel.DEBUG, 'Base curl command:');
+            this.log(ProcBot.LogLevel.DEBUG, `curl -XGET -H "Authorisation: token ${tokenDetails.token}" -H "Accept: application/vnd.github.black-cat-preview+json" https://api.github.com/`)
+        });
+    }
+
+    // Make a 'github' API call. We explicitly wrap this so that any authentication error
+    // can result in re-authentication before moving on.
+    protected gitCall = (method: any, options: any, retries?: number): Promise<any> => {
+        let badCreds = false;
+        let retriesLeft: number = retries || 3;
+
+        // We need a new Promise here, as we might need to do retries.
+        return new Promise((resolve, reject) => {
+            const runApi = () => {
+                retriesLeft -= 1;
+
+                // Run the method.
+                return method(options).catch((err: Error) => {
+                    // We only try and reauthenticate once, else we throw.
+                    if ((err.message === 'Bad credentials') && !badCreds) {
+                        badCreds = true;
+                        // Re-authenticate, then try again.
+                        return this.authenticate().then(() => {
+                            return runApi();
+                        });
+                    } else if (retriesLeft === 0) {
+                        // No more retries, just reject.
+                        reject(err);
+                    } else {
+                        // If there's more retries, try again in 5 seconds.
+                        setTimeout(() => {
+                            runApi();
+                        }, 5000);
+                    }
+                }).then((data: any) => {
+                    resolve(data);
+                });
+            };
+
+            // Kick it off.
+            runApi();
+        });
+    }
+}
+
+// Create a new GithubBot. Or rather, don't.
+export function createBot(): GithubBot {
+    throw new Error('GithubBot is not a valid instantiation of a ProcBot');
```

Comment:
If you're looking for purely virtual classes, there's an actual `abstract` keyword for that. `abstract class` will give you a class you can't directly instantiate, but can extend. I think that's what you're looking for?